File: books/understanding-computation/small_step/machine.rb
class Machine < Struct.new(:statement, :environment)

  def run
    while statement.reducible?
      log
      step
    end
    log
  end

  def step
    self.statement, self.environment = statement.reduce(environment)
  end

  def log
    puts self
  end

  def to_s
    "#{statement}, #{environment}"
  end

end

File: understanding-computation/java/transition-semantics/src/main/java/semantics/exec/Machine.java
package semantics.exec;

import semantics.env.Environment;
import semantics.stmnt.Statement;
import semantics.stmnt.StatementResult;

public class Machine {

    public Environment run(Statement statement,
                           Environment environment) {
        StatementResult result;
        while (statement.isReducible()) {
            result = statement.reduce(environment);
            statement = result.getStatement();
            environment = result.getEnvironment();
        }
        return environment;
    }

}
